datafusion/core/src/physical_plan/aggregates/priority_map.rs

```diff
@@ -0,0 +1,969 @@
+        let entry = entry.as_ref().expect("Missing node!");
+        let mut best_idx = node_idx;
+        let mut best_val = &entry.val;
+        for child_idx in left_child..=left_child + 1 {
+            if let Some(Some(child)) = self.heap.get(child_idx) {
+                if (!desc && child.val.comp(best_val) == Ordering::Greater)
+                    || (desc && child.val.comp(best_val) == Ordering::Less)
+                {
+                    best_val = &child.val;
+                    best_idx = child_idx;
+                }
+            }
+        }
+        if best_val.comp(&entry.val) != Ordering::Equal {
+            self.swap(best_idx, node_idx, mapper);
+            self.heapify_down(best_idx, mapper);
+        }
+    }
+
+    fn _tree_print(&self, idx: usize, builder: &mut ptree::TreeBuilder) -> bool {
+        let hi = self.heap.get(idx);
+        let hi = match hi {
+            None => return true,
+            Some(hi) => hi,
+        };
+        let mut valid = true;
+        if let Some(hi) = hi {
+            let label = format!("val={:?} idx={}, bucket={}", hi.val, idx, hi.map_idx);
+            builder.begin_child(label);
+            valid &= self._tree_print(idx * 2 + 1, builder); // left
+            valid &= self._tree_print(idx * 2 + 2, builder); // right
+            builder.end_child();
+            if idx != 0 {
+                let parent_idx = (idx - 1) / 2;
+                let parent = self.heap[parent_idx].as_ref().expect("Missing parent");
+                if (!self.desc && hi.val.comp(&parent.val) == Ordering::Greater)
+                    || (self.desc && hi.val.comp(&parent.val) == Ordering::Less)
+                {
+                    return false;
+                }
+            }
+        } else {
+            builder.add_empty_child("None".to_string());
+        }
+        valid
+    }
+
+    #[allow(dead_code)]
+    pub fn tree_print(&self) {
+        let mut builder = ptree::TreeBuilder::new("BinaryHeap".to_string());
+        let valid = self._tree_print(0, &mut builder);
+        let mut actual = Vec::new();
+        ptree::write_tree(&builder.build(), &mut actual).unwrap();
+        println!("{}", String::from_utf8(actual).unwrap());
+        if !valid {
+            panic!("Heap invariant violated");
+        }
+    }
+}
+
+struct CustomMap<ID: KeyType> {
+    map: RawTable<MapItem<ID>>,
+    limit: usize,
+}
+
+impl<ID: KeyType> CustomMap<ID> {
+    pub fn new(limit: usize, capacity: usize) -> Self {
+        Self {
+            map: RawTable::with_capacity(capacity),
+            limit,
+        }
+    }
+
+    pub fn find(&self, hash: u64, mut eq: impl FnMut(&ID) -> bool) -> Option<usize> {
+        let bucket = self.map.find(hash, |mi| eq(&mi.id))?;
+        let idx = unsafe { self.map.bucket_index(&bucket) };
+        Some(idx)
+    }
+
+    pub unsafe fn heap_idx_at(&self, map_idx: usize) -> usize {
+        let bucket = unsafe { self.map.bucket(map_idx) };
```

Comment:
`.bucket(index)` is the only truly unsafe operation. It is dereferencing pointers based on  the index, so the index had better be right. This is significantly faster than looking up by memoized hash, which is crucial with the index-based heap because we are constantly swapping `HeapItem`s and need to constantly update their index in the map.